### PR TITLE
docs: add hoverxref & codeautolink support. Various fixes

### LINKED ISSUE #1522: Enable hoverxref in Trio docs
docs:

https://sphinx-hoverxref.readthedocs.io/en/latest/index.html

example:

image

### PR DESCRIPTION
**Pull request by kachida**
Fixes #1522

### CODE REVIEW

**Comment by kachida** [comment]
Related issue:

https://github.com/python-trio/trio/issues/1522

**Comment by jakkdl** [comment]
Read the Docs build is failing:
```
Traceback (most recent call last):
  File "/home/docs/checkouts/readthedocs.org/user_builds/trio/envs/2968/lib/python3.11/site-packages/sphinx/registry.py", line 441, in load_extension
    mod = import_module(extname)
          ^^^^^^^^^^^^^^^^^^^^^^
  File "/home/docs/.asdf/installs/python/3.11.6/lib/python3.11/importlib/__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<frozen importlib._bootstrap>", line 1204, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1176, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1126, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "<frozen importlib._bootstrap>", line 1204, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1176, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1140, in _find_and_load_unlocked
ModuleNotFoundError: No module named 'hoverxref'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/docs/checkouts/readthedocs.org/user_builds/trio/envs/2968/lib/python3.11/site-packages/sphinx/cmd/build.py", line 285, in build_main
    app = Sphinx(args.sourcedir, args.confdir, args.outputdir,
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/docs/checkouts/readthedocs.org/user_builds/trio/envs/2968/lib/python3.11/site-packages/sphinx/application.py", line 229, in __init__
    self.setup_extension(extension)
  File "/home/docs/checkouts/readthedocs.org/user_builds/trio/envs/2968/lib/python3.11/site-packages/sphinx/application.py", line 402, in setup_extension
    self.registry.load_extension(self, extname)
  File "/home/docs/checkouts/readthedocs.org/user_builds/trio/envs/2968/lib/python3.11/site-packages/sphinx/registry.py", line 444, in load_extension
    raise ExtensionError(__('Could not import extension %s') % extname,
sphinx.errors.ExtensionError: Could not import extension hoverxref.extension (exception: No module named 'hoverxref')

Extension error:
Could not import extension hoverxref.extension (exception: No module named 'hoverxref')
```

**Comment by jakkdl** [comment]
The hoverxref seems to be working excellently, but objects in examples don't seem to be clickable (i.e. the codeautolink part)? https://trio--2968.org.readthedocs.build/en/2968/reference-testing.html

**Comment by kachida** [comment]
Thank you, @jakkdl, for adding the missing dependencies in the requirements file. I will review and confirm with you why the objects do not appear clickable.

**Comment by jakkdl** [comment]
It appears that codeautolink requires explicitly marking code blocks as python code blocks, so this does not work:
https://github.com/python-trio/trio/blob/96a5524cb1e5b4f5997f8e5638b97b145ed254de/docs/source/reference-core.rst?plain=1#L252-L258

But this does (after adding `import trio` to `codeautolink_global_preface`)
```rst
In the simplest case, you can apply a timeout to a block of code:

.. code:: python

   with trio.move_on_after(30):
       result = await do_http_get("https://...")
       print("result is", result)
   print("with block finished")
```

Skimming configuration and examples I don't find a way to configure the former to work, so will have to replace the syntax of all code blocks - but shouldn't be too bad with a regex search&replace.

File: docs/source/design.rst [hunk 1]
```diff
@@ -376,18 +378,22 @@ Specific style guidelines
   unconditionally act as cancel+schedule points.
 
 * Any function that takes a callable to run should have a signature
-  like::
+  like:
+
+  .. code:: python
 
-     def call_the_thing(fn, *args, kwonly1, kwonly2, ...)::
+     def call_the_thing(fn, *args, kwonly1, kwonly2, ...):
```
Comment (jakkdl) [inline docs/source/design.rst:385]: random typo I noticed when searching for `::$`

File: docs/source/design.rst [hunk 2]
```diff
@@ -376,18 +378,22 @@ Specific style guidelines
...
   ``kwonly2``, ... are keyword-only arguments that belong to
   ``call_the_thing``. This applies even if ``call_the_thing`` doesn't
   take any arguments of its own, i.e. in this case its signature looks
-  like::
+  like:
 
-     def call_the_thing(fn, *args)::
+  .. code:: python
+
+     def call_the_thing(fn, *args):
```
Comment (jakkdl) [inline docs/source/design.rst:396]: and a second instance of invalid python syntax

File: docs/source/tutorial.rst [hunk 3]
```diff
@@ -335,15 +345,19 @@ use ``await``. But Python's trying to keep its options open for other
 libraries that are *ahem* a little less organized about things. So
 while for our purposes we can think of ``await trio.sleep(...)`` as a
 single piece of syntax, Python thinks of it as two things: first a
-function call that returns this weird "coroutine" object::
+function call that returns this weird "coroutine" object:
+
+.. code:: pycon
```
Comment (jakkdl) [inline docs/source/tutorial.rst:350]: `pycon` is "python console"

File: docs/source/releasing.rst [hunk 4]
```diff
@@ -41,7 +41,9 @@ Things to do for releasing:
 
 * tag with vVERSION, push tag on ``python-trio/trio`` (not on your personal repository)
 
-* push to PyPI::
+* push to PyPI:
+
+  .. code::
 
     git clean -xdf   # maybe run 'git clean -xdn' first to see what it will delete
     python3 -m build
```
Comment (jakkdl) [inline docs/source/releasing.rst:49]: I don't think there is any language highlighting that's relevant here, so made it a generic code block.

File: docs/source/reference-io.rst [hunk 5]
```diff
@@ -850,19 +860,25 @@ shell doesn't provide any way to write a double quote inside a
 double-quoted string. Outside double quotes, any character (including
 a double quote) can be escaped using a leading ``^``.  But since a
 pipeline is processed by running each command in the pipeline in a
-subshell, multiple layers of escaping can be needed::
+subshell, multiple layers of escaping can be needed:
+
+.. code:: sh
```
Comment (jakkdl) [inline docs/source/reference-io.rst:865]: shell

File: docs/source/reference-io.rst [hunk 6]
```diff
@@ -850,19 +860,25 @@ shell doesn't provide any way to write a double quote inside a
...
 
 Since process creation takes a single arguments string, ``CMD.EXE``\'s
 quoting does not influence word splitting, and double quotes are not
 removed during CMD.EXE's expansion pass. Double quotes are troublesome
-because CMD.EXE handles them differently from the MSVC runtime rules; in::
+because CMD.EXE handles them differently from the MSVC runtime rules; in:
+
+.. code:: sh
 
     prog.exe "foo \"bar\" baz"
```
Comment (jakkdl) [inline docs/source/reference-io.rst:884]: there might be some windows console identifier, but I just made it `sh` /shrug

File: docs/source/conf.py [hunk 7]
```diff
@@ -181,6 +181,7 @@ def setup(app: Sphinx) -> None:
 # See https://sphinx-codeautolink.readthedocs.io/en/latest/reference.html#configuration
 codeautolink_autodoc_inject = False
 codeautolink_global_preface = """
+import trio
 from trio import *
 """
```
Comment (jakkdl) [inline docs/source/conf.py]: most of the code uses `trio.xxxx`, maybe even all of it, but not much reason not to keep `from trio import *`. Although it's plausible we should also import `testing`/`lowlevel`/etc, have not scoured the docs that thoroughly yet.

**Review by jakkdl — COMMENTED** [review]
The vast majority of the diff is near identical, so I pointed out most points where anything is different.

File: docs/source/reference-lowlevel.rst [hunk 8]
```diff
@@ -711,7 +711,8 @@ can use as a model:
 
 .. code:: python
 
-    import asyncio, trio
+    import asyncio
+    import trio
```
Comment (jakkdl) [inline docs/source/reference-lowlevel.rst:715]: Don't know why this raised errors, or by what, but it gave
```
/home/h/Git/trio/docs/source/reference-lowlevel.rst: WARNING: Could not match transformation of `asyncio` on source lines 1-1
```

**Comment by jakkdl** [comment]
I noticed one file using
```rst
.. code-block:: python3


  if bool(int('5')):
      print("hello world")
```
Turns out github (or my editor) don't recognize python3 as a language directive for highlighting the rst source code
```rst
.. code-block:: python

  if bool(int('5')):
      print("hello world")
```
(No clue why, it works in normal code blocks, and is a valid identifier both according to [linguist (used for GH code blocks)](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks#syntax-highlighting) and [pygments (used by RST)](https://pygments.org/docs/lexers/#pygments.lexers.python.PythonLexer))

After an unreasonable amount of digging I've finally figured out the difference between `.. code::` and `.. code-block::`. Sphinx [does not seem to document the existence of `.. code::` at all](https://www.sphinx-doc.org/en/master/usage/restructuredtext/basics.html#rst-directives), or [mention it as an option for showing code](https://www.sphinx-doc.org/en/master/usage/restructuredtext/directives.html#showing-code-examples), and you have to dig deep to find the docutils documentation for directives [to find it documented](https://docutils.sourceforge.io/docs/ref/rst/directives.html#code). Sphinx seems to treat them identically in ~all cases, only diverging if you want to use [options for `code-block`](https://www.sphinx-doc.org/en/master/usage/restructuredtext/directives.html#directive-code-block). So it seems like the reasonable default is to stick to `.. code-block::` in case anybody want to add any options to a block in the future.

A trailing `::` is a [literal block (sphinx)](https://www.sphinx-doc.org/en/master/usage/restructuredtext/basics.html#literal-blocks) / [preformatted block (docutils)](https://docutils.sourceforge.io/docs/user/rst/quickstart.html#preformatting-code-samples), where python highlighting was set with [`highlight_language`](https://www.sphinx-doc.org/en/master/usage/configuration.html#confval-highlight_language).

File: docs/source/reference-lowlevel.rst (hunk 8 above)
Comment (Zac-HD) [reply docs/source/reference-lowlevel.rst:715]: Let's open an upstream issue for that?  (I assume we can get a reproducer for codeautolink)

**Review by Zac-HD — APPROVED** [review]
Looks great - let's merge!

**Pull request merged**